feat: add Haskell Stack support

m0nhawk opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- To help with semantic versioning the PR title should start with one of the conventional commit types. -->
<!--- The conventional commit types for Semantic PR are: feat, fix, docs, style, refactor, perf, test, build, ci, chore, revert -->

#### Description
* Add a Haskell Stack module, when it detects `stack.yaml` file.

#### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
~Closes #~
I have a bunch of Haskell projects and I'm really missing support for Haskell in `starship`.

#### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Screenshots (if appropriate):
<img width="106" alt="Screen Shot 2019-10-16 at 12 27 15 PM" src="https://user-images.githubusercontent.com/8220926/66943255-5e13a700-f010-11e9-8e9f-420e4cdc3698.png">

#### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
- [x] I have tested using **MacOS**
- [x] I have tested using **Linux**
- [ ] I have tested using **Windows**

#### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have updated the documentation accordingly.
- [x] I have updated the tests accordingly.
- [ ] Add support for other Haskell: cabal, `.hs` files etc.

bijancn commented on src/modules/haskell.rs:
```diff
@@ -0,0 +1,60 @@
+        .is_match();
+
+    if !is_haskell_project {
+        return None;
```
there is no need to `return` here or `match` on the haskell version. You can simply say something like
```
if !is_haskell_project {
  None
} else {
  get_haskell_version().map(|haskell_version| {
  ...
  module
})
}
```

m0nhawk replied:
@bijancn That makes perfect sense. 👍

bijancn commented on src/configs/haskell.rs:
```diff
@@ -0,0 +1,23 @@
+    fn new() -> Self {
+        HaskellConfig {
+            symbol: SegmentConfig::new("λ "),
+            version: SegmentConfig::default(),
```
I still haven't worked with the new config style but why is the `version` part of the `Config` when it's not configurable?

heyrict replied:
See #472

bijancn commented on src/modules/haskell.rs:
```diff
@@ -0,0 +1,60 @@
+            module.set_style(module_style);
+
+            let formatted_version = format_haskell_version(&haskell_version)?;
+            module.new_segment("symbol", HASKELL_CHAR);
```
Is it really necessary to repeat all the default values from `src/configs/haskell.rs`?

heyrict replied:
It seems that you have mixed the old config system with the new one. Check https://github.com/starship/starship/blob/master/src/modules/ruby.rs#L26-L31 about how you can write the code.

bijancn commented on src/modules/haskell.rs:
```diff
@@ -0,0 +1,60 @@
+}
+
+fn get_haskell_version() -> Option<String> {
+    match Command::new("stack")
```
I'm not too aware of `stack` and `ghc` but should you maybe check the yaml file first for a set compiler? https://docs.haskellstack.org/en/v1.0.2/yaml_configuration/#compiler Reading from file is usually faster than starting some other command

m0nhawk replied:
The `compiler` is optional and can't be relied on.

I'm also planning to add support for `package.yaml` and `.cabal` -styled projects, then it'll make sense to first check files for compiler version and if it isn't available - run `stack` or `ghc`.

bijancn commented on src/modules/haskell.rs:
```diff
@@ -0,0 +1,60 @@
+    let mut formatted_version = String::with_capacity(version.len() + 1);
+    formatted_version.push('v');
+    formatted_version.push_str(version);
+    Some(formatted_version)
```
I don't think there is a need for a mutable String here. All you really want is to prepend a `v` here? You could format or simply add them. See also https://users.rust-lang.org/t/what-is-right-ways-to-concat-strings/3780

bijancn commented on src/modules/haskell.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+fn format_haskell_version(haskell_version: &str) -> Option<String> {
+    let formatted_version = format!("v {}", haskell_version);
```
Btw is the `v` even necessary?

m0nhawk replied:
I saw that other versions does have `v`. Though, there is no space after it, so I'm removing it.

bijancn commented:
Somehow Ubuntu and Windows Tests are still failing?

m0nhawk commented:
> Somehow Ubuntu and Windows Tests are still failing?

Same Python version issue addressed in this branch: [update-python](https://github.com/starship/starship/tree/update-python).

m0nhawk commented:
Rebased on top of `master`.

m0nhawk commented:
@bijancn I've rebased to latest `master` and all the tests are green now.

bijancn commented:
Sorry @m0nhawk I lost track of this. Looks great to me but I dont think I can merge it due to permissions. @matchai can you help?

matchai commented on docs/config/README.md:
```diff
@@ -105,6 +106,7 @@ prompt_order = [
     "ruby",
     "rust",
     "terraform",
+    "terraform",
```
It looks like this line was duplicated. 🤔 
```suggestion
```

m0nhawk replied:
Removed. 👍

matchai commented on docs/config/README.md:
```diff
@@ -605,6 +607,23 @@ The module will be shown if any of the following conditions are met:
+| `symbol`   | `"λ "`        | The symbol used before displaying the version of Haskell. |
+| `style`    | `"bold red"`  | The style for the module.                                 |
+| `disabled` | `false`       | Disables the `haskell` module.                            |
+
```
It appears you're missing the "Example" heading that other modules have:

```
### Example

```toml
# ~/.config/starship.toml

[haskell]
symbol = "🍕 "
```

m0nhawk replied:
Missed a bunch of stuff when rebasing and fixing conflicts. 🤦‍♂

matchai commented on src/modules/haskell.rs:
```diff
@@ -0,0 +1,50 @@
+}
+
+fn get_haskell_version() -> Option<String> {
+    match Command::new("stack")
```
The use of `Command::new(...)` is replaced with our `exec_command` utility function where possible:
https://github.com/starship/starship/blob/edc62f45188b742a4d20b601170ba8b9bdb5c803/src/modules/ruby.rs#L22

m0nhawk replied:
Done. 👍

m0nhawk commented:
@matchai I have added [mstksg/setup-stack](https://github.com/mstksg/setup-stack), because this requires `stack.yaml`, not only the compiler - GHC. Also, I've added few tests.

m0nhawk commented:
Also, I needed to export `$HOME` in `render_module` and `render_prompty`, because Haskell Stack requires `HOME` variable.

It would be nice to point how to better make it. 😄

chipbuster commented:
@heyrict Would you mind marking the changes you requested as resolved? (Or I can dismiss them for you if you'd like). It looks like this'll be ready to merge pretty soon and those requested changes will block the merge.

chipbuster commented:
@m0nhawk You can use `.env()` on the command from within the test--you don't have to modify the file in `common`, e.g.
```rust
  render_module("haskell")
  .use_config("blah")
  .env("HOME", var1)
```

See the test `git_repo_in_home_directory_truncate_to_repo_true` on line 443 of the directory tests for an example of how that's used.

We've also been a little skittish about using the user's actual home directory because of a slew of bugs that have come up in testing before (usually our assumptions about the state of user's `$HOME` is incorrect, which causes the tests to fail even though the module is working), so our preferred method is to make a temporary directory and then use that as the "fake home" for the tests. You can also see that in the function mentioned above.

m0nhawk commented:
@chipbuster Thank you. I've added `env()` to test with HOME directory and have a question.

Alternatively, I can make use of [`STACK_ROOT`](1).

I've also changed the workflow to include the steps to setup Haskell Stack for all OSs, but it's not running (because it's a change from the fork?).

[1]: https://docs.haskellstack.org/en/stable/GUIDE/#setting-stack-root-location

chipbuster commented:
@m0nhawk Since the test using `HOME` is `#[ignore]`d, I don't think that'll be an issue (I didn't notice that yesterday).

> and have a question.

What's that?

> ...but it's not running 

I'll try to figure out what's going on with that.

m0nhawk commented:
> What's that?
Sorry, was not very clear. It is about using `STACK_ROOT`, but probably it would complicate Github Actions setup.

> I'll try to figure out what's going on with that.
I remember that changing workflow definition in fork would result in them not running, but the only thing I can find is: [here][1]. Not sure if that's the case.

[1]: https://github.community/t5/GitHub-Actions/Run-a-GitHub-action-on-pull-request-for-PR-opened-from-a-forked/m-p/31147#M690

chipbuster commented:
Were the tests running before your latest set of commits? I seem to recall that they were, but I'm not certain.

(If they were, then that suggests that GH got confused by something that you added to the workflow YAML)

m0nhawk commented:
Yes, it was running right before the workflow change.

chipbuster commented:
Hmmm. The file looks like valid YAML, but is there any possibility that it's somehow not valid for GH Actions? Could you try reverting the YAML change and see if the tests run?

Sorry, I'm not particularly a specialist in GH Actions (I'm more familiar with Gitlab CI) and matchai doesn't see anything obviously wrong either, so we might be taking stabs in the dark here.

m0nhawk commented:
I reverted the changes and now it's running. Not sure what's wrong with it.

chipbuster commented:
Well, debugging this will be interesting, as [GitHub doesn't seem to have a YAML validator yet](https://github.community/t5/GitHub-Actions/YAML-validator-for-Github-Actions-possible-expansion-of/td-p/29557) (and the only 3rd party one I can find hasn't been updated since Actions was in early beta).

I'll try my best to take some time to play around with this later today, maybe copy the branch into my fork and try opening a (temporary) PR on that. If you figure out what happened before then, please do ping me so we can get this merged!

m0nhawk commented:
I'll also check this. I've already enabled my fork Github Actions, so I'll check if it runs there.

m0nhawk commented:
In my fork it fails with:
```
a step cannot have both the uses and run keys
```

I'll continue testing with my fork now on. Thanks for help! 😄

m0nhawk commented:
@chipbuster I managed to make it run on everything, except Windows. Will try that later locally.

m0nhawk commented:
@chipbuster Windows tests have some strange issues with environment variable, there is only `STACK_ROOT` and `LOCALAPPDATA` variables which are required for `stack` to run, but it doesn't work as expected. I'm ignoring the test for Windows for now.

chipbuster commented:
@m0nhawk Does it work on Windows at the moment?

m0nhawk commented:
Yes, `cargo test` is running fine on Windows, it's only Github Actions failing.

chipbuster commented:
@matchai Are we okay with merging this without functioning Windows CI on the module? It seems that the module and command *do* work on Windows, but we can't get GH Actions working.

matchai commented:
Yep, that's fine by me.
We'll get around to moving them over to the mocked tests soon enough 👍

chipbuster commented:
Alright, let's get this PR merged then!

(I should really get back into Haskell someday)

chipbuster commented:
@all-contributors Please add @m0nhawk for code, doc, and tests.

@m0nhawk Thanks so much for all your work on this PR, and for being patient with us through a whole bunch of changes in the project structure! We really appreciate it :smile:

m0nhawk commented:
@chipbuster Thank you (and other) for help and discussion. I found starship to be really awesome for cross-platform prompt (as a long time user of spaceship, zsh and fish). And learning some Rust was nice. :)

And, of course, started with the missing feature I need.

I will try to fix it for Windows in Github Actions, it looks like some missing environmental variables, except the one I mentioned is needed.
